File: tests/typesafety/python/exercise_user_json_solution.py
# expect: passes
# run: yes
# title: parse_user builds the User once, and the checks disappear downstream
# improves: exercise_user_json_start
import json
from dataclasses import dataclass


@dataclass(frozen=True)
class User:
    name: str
    age: int


def parse_user(body: str) -> User:
    data = json.loads(body)
    match data:
        case {"name": str(name), "age": int(age)}:
            return User(name, age)
        case _:
            raise ValueError(f"not a user: {body}")


def greeting(user: User) -> str:
    return f"hello {user.name}"


def can_vote(user: User) -> bool:
    return user.age >= 18


def summary(body: str) -> str:
    user = parse_user(body)
    return f"{greeting(user)}, can vote: {can_vote(user)}"


assert summary('{"name": "Ada", "age": 36}') == "hello Ada, can vote: True"

File: tests/typesafety/typescript/exercise_user_json_solution.ts
// expect: passes

type User = { readonly name: string; readonly age: number };

function parseUser(body: string): User {
  const data: unknown = JSON.parse(body);
  if (
    typeof data !== "object" ||
    data === null ||
    !("name" in data) ||
    !("age" in data) ||
    typeof data.name !== "string" ||
    typeof data.age !== "number"
  ) {
    throw new Error(`not a user: ${body}`);
  }
  return { name: data.name, age: data.age };
}

function greeting(user: User): string {
  return `hello ${user.name}`;
}

function canVote(user: User): boolean {
  return user.age >= 18;
}

export function summary(body: string): string {
  const user = parseUser(body);
  return `${greeting(user)}, can vote: ${canVote(user)}`;
}
